### ISSUE TITLE
Can I have multi-key discriminators?

### ISSUE THREAD

**Issue by justinmchase** [Automattic/mongoose#5846]
I want to be able to discriminate on multiple keys but can't quite figure out how to make it work. For example what would be nice:

```js
const schema = new Schema({ ... }, { discriminatorKeys: ['type', 'state'] })
const Order = model('Order', schema)

Example.discriminator({ type: 'Online', state: 'pending' }, new Schema(...))
Example.discriminator({ type: 'Online', state: 'submitted' }, new Schema(...))

Order.create({ type: 'Online', state: 'pending', ... }, (err, res) => ...)
```

One thing I tried which didn't seem to work was to make virtual sets which would update the real discriminator key:
```js
const schema = new Schema({ __t: String }, { discriminatorKey: '__t' })
schema.virtual('type').set(function (val) {
  this._type = val
  if (this._state && this._type) this.__t = `${this._type}_${this._state}`
})
schema.virtual('state').set(function (val) {
  this._state = val
  if (this._state && this._type) this.__t = `${this._type}_${this._state}`
})
```

Is there any decent way to handle the setting of the discriminator key virtually or any hooks or anything? I tried setting `schema.path('__t').get / set` also and no dice.

**EDIT: rephrasing whole question**

**Comment by vkarpov15** [comment]
That's an interesting idea, what's your use case? Right now there is no way to do that because `discriminatorKey` is always singular, we do not support a `discriminatorKeys` option.

**Comment by justinmchase** [comment]
Sorry about the slow response.

The scenario I had was simply that I have a case where the object has standard discriminator model inheritance but I also wanted to differentiate between states of the object and apply different schemas.

So essentially I wanted to use discriminators for a state machine but I also needed to discriminate based on types. So I thought if I could discriminate across two fields then I could solve this.

A simple example would be a type of form where a user can submit a claim. There are several different kinds of claims a user can submit. The discriminator on the `type` of the Claim works as intended for this.

However when the claim is first created the claim enters the 'unsubmitted' state. Based on the Claim `type` the claim may have a variety of fields but when its unsubmitted they may be null. As the user gathers information over time they may enter data into the fields. Once completed, the user may `submit()` the form which would switch it to the `'submitted'` state, and would switch to the submitted schema where the various properties are required. If a field is not filled out a validation error would occur and the change to the submitted state would not be persisted.

This was my intention when I created this issue. I realize there is another issue that would need to be resolved in order for this to solve my problem as intended, which would be if a discriminator fields value changed then I would need it to also swap the schema to a new discriminator type. That appears to be unsupported.

I managed to solve this by using the discriminator for the simple type as it was desinged to be used. Then I created a plugin which accepts a statemachine description as a parameter and it dynamically validates properties based on the state of the object and the state machine definition.

Its similar to the way this one works [mongoose-statemachine](https://www.npmjs.com/package/mongoose-statemachine) but in addition to validating state transitions I allowed individual rules to be applicable to properties but only in certain states. I also added the ability to requires a user to have certain permissions to allow the transition.

Anyway, I had to sort of wing it but I would be interested in any feedback about the "proper" way to solve some of these situations to make sure I'm not too far off the beaten trail. I am a little surprised that state machine based validation isn't more prevalent, so I'm wondering if I'm just doing it wrong.

Thanks for the reply.
